### ISSUE TITLE
Parsing delays on comments

### ISSUE THREAD

**Issue by ThunderFrame** [rubberduck-vba/Rubberduck#1502]
I was going to test some refactorings when the module is approaching the maximum permitted number of lines, but I ran into this problem when setting up the module...

New Excel 2016 workbook. Add a standard module, and add these lines:

```vb
'Line 1
'Line 2
'Line 3
...
'Line 65528
'Line 65529
'Line 65530

```

So the module has 65,530 lines, and they're *all* comments. I had expected Rubberduck to be very efficient with parsing comments, and even more efficient with inspecting comments.

Rubberduck shows Parsing for **105 seconds**
Rubberduck shows Inspecting for **120 seconds**
Total Refresh times is **3 minutes 45 seconds, which seems high, for 0 lines of executable VBA code.**

PC is i5-6500U @2.3Ghz and 8GB ram.

Am I correct in thinking that the only interesting thing about comments is that they might contain an annotation?

**Comment by Vogel612** [comment]
That assessment is correct. Unfortunately Annotations are allowed anywhere within the comment, so basically each single character has to be checked against the Annotation Rule. This makes for a rather inefficient parse-run because every time we don't match an annotation, we have to backtrack, which is expensive.

It shouldn't take nearly that long to run Inspections, though... That's what I find more interesting

**Comment by ThunderFrame** [comment]
hmmm, I wonder if we could force the first annotation to be at the start of the comment, that would increase the efficiency, and allow a user to easily suppress an annotation by prefixing the annotation with *any*/*certain* characters..

For example, the second line would *not* be recognised as an annotation and would disable the folder assignment:
```
'@Folder Foo
`-@Folder Foo
```

**Comment by ThunderFrame** [comment]
I just tried this again, but have 65000 blank lines followed by:

```
Sub foo()
  Beep
End Sub
```

The parse is taking considerably *longer* (currently 5 minutes and still running, or possibly it's stuck on Parsing?)

**Comment by retailcoder** [comment]
I'm not marking this one as a bug, because it's really just a performance edge case, but it's good to know that comments and empty lines are a parsing bottleneck.

**Comment by autoboosh** [comment]
1. It seems that ANTLR4 doesn't actually use backtracking but their own algorithm which is based on launching "subparsers at decision points in lockstep with a shared stack", the one with a match wins, if there are multiple "matches", the one with the lower production rule wins. Here's the paper describing their algorithm: http://www.antlr.org/papers/allstar-techreport.pdf
2. And I just wanted to make this clear so that there are no misunderstandings. We don't actually allow annotations to appear within a comment. The comment has to start with an annotation to be treated as such.

There's a common performance improvement suggestion if you search on the Internet which is using SLL prediction mode instead of LL. Parsing the 65k comments with SLL mode makes it 20 times faster (it then parses in 500ms for me, that 10x to get speeds on a "normal" PC I suppose?). The differences between SLL and LL are also described in ANTLR4 paper. Here's a blog post with some interesting comments on ANTLR4 performance: https://groups.google.com/forum/#!topic/antlr-discussion/aAyGkDU18IA.

The problem with switching to SLL is that we might have to rewrite the grammar because it's "stricter" from what I understand. I've tried parsing a module with normal code and it fails and falls back to LL. But if we could rewrite the grammar to be 100% SLL we might see a noticeable performance improvement everywhere. The ANTLR4 paper even says that their SLL mode is 8x faster when parsing 100M+ java source code.

I'll try to switch the precompiler plus expression parser to the SLL/LL fallback approach and see how it goes and then try to find out why the main grammar fails.

I can't comment on the part that comes after parsing though.

**Comment by autoboosh** [comment]
The parser should be a lot faster now. The comments example mentioned in the first post takes about 500ms to parse here now. @ThunderFrame (or anyone really): Can you use the latest version and tell me how long it takes?

**Comment by autoboosh** [comment]
Fixed with #1560.
